File: src/parser.py
"""
parser.py — LLM-based Resume Parser

Uses a local GGUF model (Phi-3 Mini / Qwen2.5) via llama-cpp-python
to parse raw resume text into structured JSON fields.

Falls back to regex extraction if LLM output is not valid JSON.
"""

import json
import re
import os
import logging
from typing import Optional

logger = logging.getLogger(__name__)

# Global model cache — loaded once per session
_llm_instance = None

PROMPT_TEMPLATE = """You are a resume parser. Extract the following fields from the resume text below and return ONLY valid JSON — no explanation, no markdown code blocks, no extra text.

Fields to extract:
- name: string (full name of the candidate)
- email: string (email address)
- phone: string (phone number)
- skills: list of strings (technical and soft skills)
- education: list of objects with keys: degree (string), college (string)
- experience: list of objects with keys: company (string), role (string), years (number)
- projects: list of strings (project names or descriptions)

Resume Text:
{resume_text}

Return ONLY valid JSON:"""


def load_model(model_path: str):
    """
    Load and cache the GGUF model using llama-cpp-python.

    Args:
        model_path: Absolute path to the .gguf model file.

    Returns:
        Llama model instance.

    Raises:
        FileNotFoundError: If the model file does not exist.
    """
    global _llm_instance

    if _llm_instance is not None:
        return _llm_instance

    if not os.path.exists(model_path):
        raise FileNotFoundError(
            f"GGUF model not found at: {model_path}\n"
            "Please download Phi-3 Mini from:\n"
            "https://huggingface.co/microsoft/Phi-3-mini-4k-instruct-gguf"
        )

    try:
        from llama_cpp import Llama
        _llm_instance = Llama(
            model_path=model_path,
            n_ctx=4096,
            n_threads=os.cpu_count(),
            verbose=False,
        )
        logger.info(f"GGUF model loaded from: {model_path}")
        return _llm_instance
    except ImportError:
        raise ImportError(
            "llama-cpp-python is not installed. Run: pip install llama-cpp-python"
        )


def parse_resume(raw_text: str, model) -> dict:
    """
    Parse raw resume text into structured JSON using the local LLM.

    Args:
        raw_text: Extracted text from a resume file.
        model: Loaded Llama model instance.

    Returns:
        Dictionary with parsed candidate fields.
        Falls back to regex extraction if JSON parsing fails.
    """
    # Truncate very long resumes to avoid context overflow
    max_chars = 6000
    if len(raw_text) > max_chars:
        raw_text = raw_text[:max_chars] + "\n[truncated]"

    prompt = PROMPT_TEMPLATE.format(resume_text=raw_text)

    try:
        response = model(
            prompt,
            max_tokens=1024,
            temperature=0.1,
            top_p=0.95,
            stop=["```", "\n\n\n"],
        )
        output_text = response["choices"][0]["text"].strip()

        # Strip markdown code fences if present
        output_text = re.sub(r"^```json?\s*", "", output_text)
        output_text = re.sub(r"\s*```$", "", output_text)

        parsed = json.loads(output_text)
        logger.info("LLM JSON parsing successful")
        return parsed

    except (json.JSONDecodeError, KeyError, Exception) as e:
        logger.warning(f"LLM JSON parsing failed: {e}. Falling back to regex.")
        return _regex_fallback(raw_text)


def _regex_fallback(raw_text: str) -> dict:
    """
    Regex-based fallback extractor for key resume fields.

    Args:
        raw_text: Raw resume text.

    Returns:
        Partially populated candidate dict from regex patterns.
    """
    result = {
        "name": "",
        "email": "",
        "phone": "",
        "skills": [],
        "education": [],
        "experience": [],
        "projects": [],
    }

    # Email
    email_match = re.search(r"[\w.+-]+@[\w-]+\.[a-z]{2,}", raw_text, re.I)
    if email_match:
        result["email"] = email_match.group(0)

    # Phone
    phone_match = re.search(r"(\+?\d[\d\s\-().]{8,14}\d)", raw_text)
    if phone_match:
        result["phone"] = phone_match.group(0).strip()

    # Name (first non-empty line, assumed to be name)
    lines = [line.strip() for line in raw_text.split("\n") if line.strip()]
    if lines:
        result["name"] = lines[0]

    # Skills (look for common tech keywords)
    KNOWN_SKILLS = [
        "Python", "JavaScript", "TypeScript", "Java", "C++", "C#", "Go", "Rust",
        "React", "Angular", "Vue", "Node.js", "Django", "Flask", "FastAPI",
        "SQL", "MongoDB", "PostgreSQL", "MySQL", "SQLite", "Redis",
        "Docker", "Kubernetes", "AWS", "Azure", "GCP", "Linux",
        "Machine Learning", "Deep Learning", "NLP", "TensorFlow", "PyTorch",
        "Git", "REST", "GraphQL", "HTML", "CSS",
    ]
    found_skills = [skill for skill in KNOWN_SKILLS if skill.lower() in raw_text.lower()]
    result["skills"] = found_skills

    return result

File: js/parser.js
/**
 * Client-side resume parser.
 * Extracts structured fields from raw resume text using regex heuristics.
 */

const SKILL_KEYWORDS = [
  'python', 'java', 'javascript', 'typescript', 'c++', 'c#', 'go', 'rust', 'ruby', 'php',
  'swift', 'kotlin', 'scala', 'r', 'matlab', 'perl', 'sql', 'nosql', 'html', 'css',
  'react', 'angular', 'vue', 'svelte', 'next.js', 'node.js', 'express', 'django', 'flask',
  'fastapi', 'spring', 'spring boot', '.net', 'laravel', 'rails',
  'tensorflow', 'pytorch', 'keras', 'scikit-learn', 'pandas', 'numpy', 'opencv',
  'machine learning', 'deep learning', 'nlp', 'computer vision', 'ai',
  'aws', 'azure', 'gcp', 'docker', 'kubernetes', 'terraform', 'ansible',
  'jenkins', 'github actions', 'gitlab ci', 'ci/cd', 'devops',
  'git', 'linux', 'bash', 'powershell', 'rest api', 'graphql', 'grpc',
  'mongodb', 'postgresql', 'mysql', 'redis', 'elasticsearch', 'dynamodb', 'sqlite',
  'kafka', 'rabbitmq', 'nginx', 'apache',
  'figma', 'sketch', 'adobe xd', 'photoshop', 'illustrator',
  'agile', 'scrum', 'jira', 'confluence', 'trello',
  'data analysis', 'data engineering', 'data science', 'etl', 'power bi', 'tableau',
  'blockchain', 'solidity', 'web3',
  'cybersecurity', 'penetration testing', 'networking',
  'android', 'ios', 'flutter', 'react native', 'xamarin',
  'selenium', 'cypress', 'jest', 'pytest', 'junit', 'mocha',
  'streamlit', 'gradio', 'hugging face', 'transformers', 'langchain',
  'microservices', 'system design', 'design patterns', 'oop',
  'communication', 'leadership', 'teamwork', 'problem solving', 'critical thinking'
];

function extractName(text) {
  const lines = text.split('\n').map(l => l.trim()).filter(l => l.length > 0);
  // First non-empty line is usually the name
  for (const line of lines.slice(0, 5)) {
    // Skip lines that look like headers or contain email/phone
    if (line.match(/@/) || line.match(/\d{10}/) || line.length > 60) continue;
    if (line.match(/resume|curriculum|vitae|cv|objective|summary/i)) continue;
    // A name is typically 2-4 words, all starting with uppercase
    const words = line.split(/\s+/);
    if (words.length >= 2 && words.length <= 5) {
      const looksLikeName = words.every(w => w.match(/^[A-Z]/));
      if (looksLikeName) return line;
    }
  }
  // Fallback: first short line
  for (const line of lines.slice(0, 3)) {
    if (line.length > 3 && line.length < 50 && !line.match(/@|\d{5,}/)) return line;
  }
  return 'Unknown';
}

function extractEmail(text) {
  const match = text.match(/[a-zA-Z0-9._%+\-]+@[a-zA-Z0-9.\-]+\.[a-zA-Z]{2,}/);
  return match ? match[0] : '';
}

function extractPhone(text) {
  const patterns = [
    /(?:\+?\d{1,3}[-.\s]?)?\(?\d{3}\)?[-.\s]?\d{3}[-.\s]?\d{4}/,
    /(?:\+?\d{1,3}[-.\s]?)?\d{10}/,
    /(?:\+?\d{1,3}[-.\s]?)?\d{5}[-.\s]?\d{5}/
  ];
  for (const p of patterns) {
    const match = text.match(p);
    if (match) return match[0];
  }
  return '';
}

function extractSkills(text) {
  const lower = text.toLowerCase();
  const found = [];
  for (const skill of SKILL_KEYWORDS) {
    // Match whole-word (with some flexibility for special chars)
    const escaped = skill.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
    const regex = new RegExp(`\\b${escaped}\\b`, 'i');
    if (regex.test(lower)) {
      found.push(skill.charAt(0).toUpperCase() + skill.slice(1));
    }
  }
  return [...new Set(found)];
}

function extractEducation(text) {
  const education = [];
  const degreePatterns = [
    /(?:B\.?(?:Tech|Sc|A|E|Com|Eng)|M\.?(?:Tech|Sc|A|BA|Com|Eng)|Ph\.?D|MBA|BCA|MCA|B\.?S|M\.?S|Bachelor|Master|Diploma|Associate)/gi
  ];
  const lines = text.split('\n');
  for (let i = 0; i < lines.length; i++) {
    const line = lines[i];
    for (const pattern of degreePatterns) {
      pattern.lastIndex = 0;
      const match = pattern.exec(line);
      if (match) {
        // Try to find college name nearby
        const context = lines.slice(Math.max(0, i - 1), i + 3).join(' ');
        const collegePat = /(?:university|institute|college|school|iit|nit|iiit|bits|vit|srm|anna|amity|manipal|jntu|osmania|delhi|mumbai|bangalore|chennai|pune|hyderabad)\b[^,.\n]*/i;
        const collegeMatch = context.match(collegePat);
        education.push({
          degree: match[0],
          college: collegeMatch ? collegeMatch[0].trim() : ''
        });
        break;
      }
    }
  }
  return education;
}

function extractExperience(text) {
  const experience = [];
  const lines = text.split('\n');
  // Look for patterns like "Company Name | Role | Duration" or similar
  const expPattern = /(?:(\d+(?:\.\d+)?)\s*(?:\+?\s*)?(?:years?|yrs?))/gi;
  const roleKeywords = /(?:engineer|developer|architect|analyst|manager|lead|intern|designer|consultant|specialist|administrator|director|coordinator|associate|scientist|researcher)/i;

  for (let i = 0; i < lines.length; i++) {
    const line = lines[i].trim();
    if (roleKeywords.test(line)) {
      const yearsMatch = lines.slice(i, i + 3).join(' ').match(expPattern);
      const years = yearsMatch ? parseFloat(yearsMatch[0]) : 0;
      // Try to extract company from the line or surrounding lines
      const parts = line.split(/[|•·–—,]/);
      experience.push({
        role: parts[0]?.trim() || line.substring(0, 50),
        company: parts[1]?.trim() || '',
        years: years
      });
    }
  }
  return experience.slice(0, 10); // Cap at 10
}

function extractProjects(text) {
  const projects = [];
  const lines = text.split('\n');
  let inProjectSection = false;

  for (const line of lines) {
    const trimmed = line.trim();
    if (trimmed.match(/^projects?\b/i)) {
      inProjectSection = true;
      continue;
    }
    if (inProjectSection) {
      if (trimmed.match(/^(education|experience|skills|certif|achiev|award|hobby|interest|referenc)/i)) {
        break;
      }
      if (trimmed.length > 5 && trimmed.length < 120) {
        const cleaned = trimmed.replace(/^[•\-▸▪○●◦‣]\s*/, '');
        if (cleaned.length > 3) projects.push(cleaned);
      }
    }
  }
  return projects.slice(0, 8);
}

function parseResume(text) {
  return {
    name: extractName(text),
    email: extractEmail(text),
    phone: extractPhone(text),
    skills: extractSkills(text),
    education: extractEducation(text),
    experience: extractExperience(text),
    projects: extractProjects(text),
    resumeText: text
  };
}

// Export
window.ResumeParser = { parseResume };
